Added From<Vec<NonZeroU8>> for CString

danielhenrymantilla opened the pull request:
Added a `From<Vec<NonZeroU8>>` `impl` for `CString`

# Rationale

  - `CString::from_vec_unchecked` is a subtle function, that makes `unsafe` code harder to audit when the generated `Vec`'s creation is non-trivial. This `impl` allows to write safer `unsafe` code thanks to the very explicit semantics of the `Vec<NonZeroU8>` type.

  - One such situation is when trying to `.read()` a `CString`, see issue #59229.

      - this lead to a PR: #59314, that was closed for being too specific / narrow (it only targetted being able to `.read()` a `CString`, when this pattern could have been generalized).

     - the issue suggested another route, based on `From<Vec<NonZeroU8>>`, which is indeed a less general and more concise code pattern.

  - quoting @shnatsel:

      - >  For me the main thing about making this safe is simplifying auditing - people have spent like an hour looking at just this one unsafe block in libflate because it's not clear what exactly is unchecked, so you have to look it up when auditing anyway. This has distracted us from much more serious memory safety issues the library had.
Having this trivial impl in stdlib would turn this into safe code with compiler more or less guaranteeing that it's fine, and save anyone auditing the code a whole lot of time.

rust-highfive commented:
r? @withoutboats

(rust_highfive has picked a reviewer for you, use r? to override)

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
     }
 }
 
+#[unstable(feature = "cstring_from_vec_of_nonzerou8", issue = "64070")]
```
Implementations are insta-stable.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=7169&view=logs&jobId=4f922444-fdfe-5dcf-b824-02f86439ef14) ([raw log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/7169/logs/45)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-09-01T17:30:24.9862540Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-09-01T17:30:25.6788628Z ##[command]git config gc.auto 0
2019-09-01T17:30:25.6791906Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-09-01T17:30:25.6795190Z ##[command]git config --get-all http.proxy
2019-09-01T17:30:25.6799980Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/64069/merge:refs/remotes/pull/64069/merge
---
2019-09-01T17:36:38.0473105Z     Checking backtrace v0.3.35
2019-09-01T17:36:43.2265818Z error: impl has missing stability attribute
2019-09-01T17:36:43.2267576Z    --> src/libstd/ffi/c_str.rs:722:1
2019-09-01T17:36:43.2268155Z     |
2019-09-01T17:36:43.2268657Z 722 | / impl From<Vec<NonZeroU8>> for CString {
2019-09-01T17:36:43.2269186Z 723 | |     /// Converts a [`Vec`]`<`[`NonZeroU8`]`>` into a [`CString`] without
2019-09-01T17:36:43.2269717Z 724 | |     /// copying nor checking for inner null bytes.
2019-09-01T17:36:43.2270910Z ...   |
2019-09-01T17:36:43.2271338Z 793 | |     }
2019-09-01T17:36:43.2271800Z 794 | | }
2019-09-01T17:36:43.2272242Z     | |_^
---
2019-09-01T17:36:43.4492110Z == clock drift check ==
2019-09-01T17:36:43.4512422Z   local time: Sun Sep  1 17:36:43 UTC 2019
2019-09-01T17:36:43.6012744Z   network time: Sun, 01 Sep 2019 17:36:43 GMT
2019-09-01T17:36:43.6017002Z == end clock drift check ==
2019-09-01T17:36:50.5865241Z ##[error]Bash exited with code '1'.
2019-09-01T17:36:50.5899521Z ##[section]Starting: Checkout
2019-09-01T17:36:50.5902055Z ==============================================================================
2019-09-01T17:36:50.5902124Z Task         : Get sources
2019-09-01T17:36:50.5902167Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

clarfon commented:
Perhaps there could be a method to return the nonzero bytes as a slice as well?

Shnatsel commented:
> Perhaps there could be a method to return the nonzero bytes as a slice as well?

Sounds reasonable to me. That should be its own PR though, as that can be feature-gated for some time while trait implementations are instantly stable.

JohnCSimon commented:
Ping from triage
@Centril @withoutboats This PR is still waiting on review. 
CC @danielhenrymantilla 

Thanks.

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+        ///   - for any `cap: usize`, `Layout<[T; cap]>` needs to be equal to
+        ///     `Layout<[U; cap]>` (for the allocator)
+        #[inline]
+        unsafe
```
what's with the line break here?

danielhenrymantilla replied:
Woops, my own Rust style leaked here (Out of topic: I see `pub` and `unsafe`, among others, as _markers_, hence the logic of them being on an extra line rather than on the same line).

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+        #[inline]
+        unsafe
+        fn transmute_vec<T, U>(v: Vec<T>) -> Vec<U> {
+            // necessary conditions for `Layout<[T; N]> == Layout<[U; N]>`
```
Why not check `Layout::array::<T / U>().unwrap()` against each other?

Centril replied:
```suggestion
            // Necessary conditions for `Layout<[T; N]> == Layout<[U; N]>`:
```

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+        ///     `[T; length]` and `[U; length]`.
+        ///
+        ///   - for any `cap: usize`, `Layout<[T; cap]>` needs to be equal to
+        ///     `Layout<[U; cap]>` (for the allocator)
```
```suggestion
        ///     `Layout<[U; cap]>` (for the allocator).
```

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+            //
+            //   - this is the pattern blessed by transmute's documentation
+            //     (and replacing `mem::forget` by `ManuallyDrop`, _c.f._,
+            //     issue #62553)
```
```suggestion
            //     issue #62553).
```

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+            //     issue #62553)
+            //
+            //   - `transmute_vec` contract guarantees sound transmuting
+            //     between `[T; length]` and `[U; length]`
```
```suggestion
            //     between `[T; length]` and `[U; length]`.
```

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+            let length: usize = v.len();
+            let capacity: usize = v.capacity();
+            let ptr: *mut T = {
+                // end-of-scope `mem::forget` but without aliasing problems
```
```suggestion
                // End-of-scope `mem::forget` but without aliasing problems.
```

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+                // end-of-scope `mem::forget` but without aliasing problems
+                let mut v = mem::ManuallyDrop::<Vec<T>>::new(v);
+                let v: &mut Vec<T> = &mut *v;
+                v.as_mut_ptr() // cannot be aliased
```
```suggestion
                v.as_mut_ptr() // Cannot be aliased.
```

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+            //   - `v` cannot contain null bytes, given the type-level
+            //     invariant of `NonZeroU8` (this would still apply even if
+            //     this invariant was a safety invariant and not a validity
+            //     invariant)
```
```suggestion
            //     invariant).
```

danielhenrymantilla replied:
Always on point ;)

Centril commented:
r? @RalfJung for checking the safety argument more in-depth. Also cc @rkruppe.

rkruppe commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+            //
+            //   - `v` cannot contain null bytes, given the type-level
+            //     invariant of `NonZeroU8` (this would still apply even if
+            //     this invariant was a safety invariant and not a validity
```
Nit: strike the hypothetical. I am reasonably sure you do need the safety invariant here. Most operations on vectors don't assert validity of any elements except perhaps the ones they move into or out of the vec. Of course, a `Vec<T>` is only *safe* if elements `0..len` are safe at `T`, so the soundness argument can rest on that plus the safety/validity (same thing) of the `NonZeroU8`.

(Insert standard disclaimer about all of these details being not yet ratified here.)

danielhenrymantilla replied:
You're right, I will remove the part between brackets then (I'm still glad I put it, it is good to clearly think about these things before merging).

rkruppe commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
+            Vec::<U>::from_raw_parts(ptr as *mut U, length, capacity)
+        }
+
+        let v = unsafe {
```
A type annotation won't hurt here IMO.

danielhenrymantilla replied:
Agreed

danielhenrymantilla commented on src/libstd/ffi/c_str.rs:
```diff
@@ -739,17 +739,16 @@ impl From<Vec<NonZeroU8>> for CString {
+                crate::alloc::Layout::new::<U>(),
+            );
+            // The previous assert should imply the following one:
+            debug_assert_eq!(
```
The first check is an assert since it is expected to be easily optimized out (only const / compile-time information here).
On the other hand, the check involving `Layout::array` cannot be `const`, since it involves the runtime value of the vector capacity.
That being said, I expect `Layout<T> == Layout<U>` to imply that for all `N: usize`, `Layout<[T; N]> == Layout<[U; N]>`.
If this were not to be correct / guaranteed by the language, this debug assertion could be upgraded to a normal `assert!`; in this case however, the function is only used with the monomorphisation to `T = NonZeroU8` and `V = u8`, which does not require such check at compile-time.

RalfJung replied:
I think it would be a serious bug in `Layout::array` if this is not given... its comparison test would fail to take something into account, or so.

Not sure if this check is worth the clutter it introduces.

danielhenrymantilla replied:
Great! I will remove the `debug_assert!` involving the `Layout::array` check

RalfJung commented:
I'm afraid my Rust inbox is swamped, I won't be able to review this -- I can take a glance, but don't have time to do a full review.

Any takers?

(Also is there some way to tell bors "sorry, not me", if I don't know whom else to assign?)

RalfJung commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,77 @@ impl From<Box<CStr>> for CString {
+                // End-of-scope `mem::forget` but without aliasing problems.
+                let mut v = mem::ManuallyDrop::<Vec<T>>::new(v);
+                let v: &mut Vec<T> = &mut *v;
+                v.as_mut_ptr() // Cannot be aliased.
```
`ManuallyDrop` implements `Deref`; why is the intermediate `let v: &mut Vec<T> = ...` needed?

danielhenrymantilla replied:
It's just me being (overly?) cautious with type inference / method resolution within `unsafe` code , so I wanted to have that `Deref` be very explicit.

Mark-Simulacrum commented:
Going to re-assign to @rkruppe as they appear to have done some review here but feel free to reassign to someone else if you don't feel confident reviewing this.

rkruppe commented:
I'm fine with reviewing this. Will take another close look later for due digilence but from what I recall this should be in good shape (modulo removing that `debug_assert` that Ralf pointed out is redundant).

I think some @rust-lang/libs sign-off is also required since this is a new insta-stable API addition.

Mark-Simulacrum commented:
I've nominated to get someone from T-libs to kick off an FCP merge here

SimonSapin commented:
Is there any precedent for working with `Vec<NonZeroU8>`? How would one go about constructing a value of that type?

rkruppe commented:
r=me pending T-libs approval

@SimonSapin I'm not aware of any existing APIs for that type specifically, so no precedent, but the motivating example from libflate in https://github.com/rust-lang/rust/issues/59229 can be modified to build up a `Vec<NonZeroU8>` and convert it to a CString once a 0 byte is read.

SimonSapin commented:
I’m not disputing that the conversion is correct, but that it is useful. The code in #59229 that would use this conversion doesn’t look better than the original to me.

(To answer my own question: the way to construct `Vec<NonZeroU8>` in #59229 is one byte at a time with `NonZeroU8::new`.)

rkruppe commented:
The advantage of using this API over the code in #59229 is that the `unsafe` is isolated in libstd. So it protects against mistakes such as adding the 0 terminator to the `Vec<u8>` before calling `CString::from_vec_unchecked` (happened in https://github.com/rust-lang/rust/issues/59229#issuecomment-515399339) and simplifies auditing for memory safety. The latter point was already made in the PR description with this quote from @Shnatsel:

> For me the main thing about making this safe is simplifying auditing - people have spent like an hour looking at just this one unsafe block in libflate because it's not clear what exactly is unchecked, so you have to look it up when auditing anyway. This has distracted us from much more serious memory safety issues the library had.
Having this trivial impl in stdlib would turn this into safe code with compiler more or less guaranteeing that it's fine, and save anyone auditing the code a whole lot of time.

I'm in no position to argue whether it is useful for other crates but the libflate snippet in question does benefits from this API addition IMO.

Shnatsel commented:
The primary use case motivating this is constructing a `CString` from a reader. There is currently no safe way to do that without scanning the `CString` for null bytes twice.

A bespoke unsafe implementation works, but without an std abstraction people need to reinvent it every time, and get it right. Even if done right, presence of a bespoke unsafe block increases the cost of auditing safety of the crate.

This API addition isolates the unsafety inside std, so it can be implemented correctly once and for all.

JohnTitor commented:
Marked as `waiting-on-team` because of waiting T-libs approval

Centril commented:
r? @sfackler

Centril commented on src/libstd/ffi/c_str.rs:
```diff
@@ -718,6 +719,72 @@ impl From<Box<CStr>> for CString {
     }
 }
 
+#[stable(feature = "cstring_from_vec_of_nonzerou8", since = "1.40.0")]
```
```suggestion
#[stable(feature = "cstring_from_vec_of_nonzerou8", since = "1.41.0")]
```

JohnTitor commented:
Ping from triage: @sfackler and @rust-lang/libs could you review this PR?

sfackler commented:
@rfcbot fcp merge

dtolnay commented:
I am on board with adding the impl, but the implementation here looks nuts to me. Is this the way we expect unsafe code in the standard library to be written? Is this the way we recommend unsafe code outside the standard library to be written?

I would much rather see something like:

```rust
fn from(vec_nonzero: Vec<NonZeroU8>) -> CString {
    // Safety: ...

    let mut vec_nonzero = ManuallyDrop::new(vec_nonzero);
    let ptr: *mut NonZeroU8 = vec_nonzero.as_mut_ptr();
    let len = vec_nonzero.len();
    let cap = vec_nonzero.capacity();

    unsafe {
        let vec_u8 = Vec::from_raw_parts(ptr as *mut u8, len, cap);
        CString::from_vec_unchecked(vec_u8)
    }
}
```

where the code fits easily on one screen and the safety reasoning is explained all at once.

sfackler commented:
Yeah, this should get cleaned up.

@bors r-

SimonSapin commented:
Although that landed more recently than this PR was opened, now it can use `let (ptr, len, cap) = v.into_raw_parts();` https://github.com/rust-lang/rust/issues/65816

danielhenrymantilla commented:
This PR and its very verbose code is kind of what triggered the discussions about transmuting `Vec`s and the added `Vec::into_raw_parts` (and hopefully `Vec::transmute` or a more generalized variant).

The main thing bloating the code is thus the "soundly transmute a `Vec`" thing, which was factored into its own function as an example.

Now, with `Vec::into_raw_parts`, the implementation can be simplified, which I will happily do. 

I disagree, however, into factoring both `unsafe` block into one: the more fine grained the `unsafe` blocks can be, the more resilient the code will be to further changes.

In this case, for instance, there are _two_ `unsafe` things going on:

  - transmuting `Vec<NonZeroU8>` into a `Vec<u8>` (valid thanks to `transmute<NonZeroU8, u8>` being always sound AND thanks to `Layout<NonZeroU8> == Layout<u8>`),

  - and then using the unchecked constructor of `CString`.

---

As an example of how delicate this may be, @dtolnay's example without `into_raw_parts` could be considered to be violating Rust aliasing rules (when `mem::forget`ting `vec_nonzero`, it gives ownership of the `Vec` into a function, and thus asserts that the pointee is at that point unaliased, invalidating `ptr`).

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14292&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14292/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14292/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-24T10:50:46.0052387Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-24T10:50:46.0174410Z ##[command]git config gc.auto 0
2019-11-24T10:50:46.0176899Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-24T10:50:46.0178931Z ##[command]git config --get-all http.proxy
2019-11-24T10:50:47.0018209Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/64069/merge:refs/remotes/pull/64069/merge
---
2019-11-24T10:57:09.9272732Z    Compiling panic_abort v0.0.0 (/checkout/src/libpanic_abort)
2019-11-24T10:57:10.1196925Z    Compiling backtrace v0.3.40
2019-11-24T10:57:11.0329609Z    Compiling rustc-std-workspace-alloc v1.99.0 (/checkout/src/tools/rustc-std-workspace-alloc)
2019-11-24T10:57:11.1416434Z    Compiling panic_unwind v0.0.0 (/checkout/src/libpanic_unwind)
2019-11-24T10:57:13.7666936Z error[E0658]: use of unstable library feature 'vec_into_raw_parts': new API
2019-11-24T10:57:13.7667635Z     |
2019-11-24T10:57:13.7667635Z     |
2019-11-24T10:57:13.7668012Z 764 |             let (ptr, len, cap): (*mut NonZeroU8, _, _) = Vec::into_raw_parts(v);
2019-11-24T10:57:13.7668689Z     |
2019-11-24T10:57:13.7668689Z     |
2019-11-24T10:57:13.7669149Z     = note: for more information, see ***/issues/65816
2019-11-24T10:57:13.7669541Z     = help: add `#![feature(vec_into_raw_parts)]` to the crate attributes to enable
2019-11-24T10:57:14.8472199Z error: aborting due to previous error
2019-11-24T10:57:14.8472317Z 
2019-11-24T10:57:14.8472712Z For more information about this error, try `rustc --explain E0658`.
2019-11-24T10:57:14.8852457Z error: could not compile `std`.
---
2019-11-24T10:57:14.8943158Z   local time: Sun Nov 24 10:57:14 UTC 2019
2019-11-24T10:57:15.0395919Z   network time: Sun, 24 Nov 2019 10:57:15 GMT
2019-11-24T10:57:15.0397650Z == end clock drift check ==
2019-11-24T10:57:17.9402685Z 
2019-11-24T10:57:17.9505391Z ##[error]Bash exited with code '1'.
2019-11-24T10:57:17.9534314Z ##[section]Starting: Checkout
2019-11-24T10:57:17.9536204Z ==============================================================================
2019-11-24T10:57:17.9536270Z Task         : Get sources
2019-11-24T10:57:17.9536324Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

SimonSapin commented:
> there are _two_ `unsafe` things going on

That’s a good reason to write an explanation comment in in two parts. If feel it’s not useful to write two consecutive `unsafe {…}` blocks.

Dylan-DPC commented:
@kimundi @KodrAus @withoutboats waiting for your response on the above fcp

dtolnay commented:
I checked Kimundi's box per https://github.com/rust-lang/team/commit/bfbf48f27d09abec36aaf269c98740e261bcb129.

KodrAus commented:
Did we come to a decision on the multiple `unsafe` blocks? I don't imagine it needs to block merging, but am also of the opinion that we should combine them into one, especially now that they live alongside one another with no safe code in between. They're logically a single operation so I don't think their granularity is necessarily better here.

But thanks for working on this @danielhenrymantilla! Would you like to squash some of your commits down before merging?

danielhenrymantilla commented:
I did have a preference for the split blocks, but I seem to be at a minority here, so I have merged the `unsafe` blocks (and updated the stability annotation and squashed and rebased everything).

There should be nothing blocking the merge, now 🙂

Dylan-DPC commented:
r? @KodrAus

KodrAus commented:
@bors r+

bors commented:
:pushpin: Commit 60274a95fef57a18113f7c48be68be31ece860eb has been approved by `KodrAus`

<!-- @bors r=KodrAus 60274a95fef57a18113f7c48be68be31ece860eb -->
<!-- homu: {"type":"Approved","sha":"60274a95fef57a18113f7c48be68be31ece860eb","approver":"KodrAus"} -->

bors commented:
:hourglass: Testing commit 60274a95fef57a18113f7c48be68be31ece860eb with merge dbef35309d10bed2f09cd55b3e9ea980a58c62aa...
<!-- homu: {"type":"BuildStarted","head_sha":"60274a95fef57a18113f7c48be68be31ece860eb","merge_sha":"dbef35309d10bed2f09cd55b3e9ea980a58c62aa"} -->
